### PR TITLE
Migrations

### PR DESCRIPTION
**Pull request by Lehersun**
Configure doctrine/migrations with SQLite database

### CODE REVIEW

**Comment by Lehersun** [comment]
Problem 1

    - laminas/laminas-code is locked to version 4.4.2 and an update of this package was not requested.

    - laminas/laminas-code 4.4.2 requires php ^7.4 || ~8.0.0 -> your php version (7.3.27) does not satisfy that requirement.

  Problem 2

    - psr/log is locked to version 3.0.0 and an update of this package was not requested.

    - psr/log 3.0.0 requires php >=8.0.0 -> your php version (7.3.27) does not satisfy that requirement.

  Problem 3

    - psr/log 3.0.0 requires php >=8.0.0 -> your php version (7.3.27) does not satisfy that requirement.

    - doctrine/migrations 3.2.1 requires psr/log ^1.1.3 || ^2 || ^3 -> satisfiable by psr/log[3.0.0].

    - doctrine/migrations is locked to version 3.2.1 and an update of this package was not requested.

File: composer.json [hunk 1]
```diff
@@ -11,15 +11,19 @@
   ],
   "require-dev": {
     "phpunit/phpunit": "^9.5",
-    "squizlabs/php_codesniffer": "^3.6"
+    "squizlabs/php_codesniffer": "^3.6",
+    "doctrine/migrations": "^3.2",
+    "psr/log": "1.1.3",
+    "laminas/laminas-code": "3.4.1"
```
Comment (Magomogo) [inline composer.json]: What for this dependency?

File: composer.json [hunk 2]
```diff
@@ -11,15 +11,19 @@
...
   },
   "autoload": {
     "psr-4": {"GroupLife\\Core\\":"src/"}
   },
   "scripts": {
     "test": ["phpunit", "@composer style"],
     "style": "phpcs --standard=PSR12 src/ tests/",
-    "style-fix": "phpcbf --standard=PSR12 src/ tests/"
+    "style-fix": "phpcbf --standard=PSR12 src/ tests/",
+    "create-db": "./vendor/bin/doctrine-migrations execute --up 'GroupLife\\Migrations\\Version20210913225128'"
```
Comment (Magomogo) [inline composer.json]: We should apply all migrations, not a single one.

File: .travis.yml [hunk 3]
```diff
@@ -3,7 +3,10 @@ php:
   - '7.3'
   - '7.4'
   - '8.0'
+env:
+  - DB=sqlite
 install:
   - composer install
 script:
   - composer test
+  - composer create-db
```
Comment (Magomogo) [inline .travis.yml:11]: Hmmm, I thought we need to create a database first,  then run the tests.

File: tests/lib/Migrations/Version20210913225128.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+            CREATE TABLE `rules` (
+                id INT AUTO_INCREMENT NOT NULL, 
+                rules VARCHAR(255), 
+                schedule INT, 
+                PRIMARY KEY(id))
+                ');
+        $this->addSql('
+            ALTER TABLE `rules` ADD FOREIGNKEY schedule 
+            REFERENCES `schedules`(`id`) 
+            ON DELETE CASCADE ON UPDATE RESTRICT;');
```
Comment (Magomogo) [inline tests/lib/Migrations/Version20210913225128.php]: About naming.

Better to name tables in singular form: `schedule` and `schedule_rule`

Foreign key usually named by the connected tables: `schedule_rule_schedule_id_fk`

File: tests/lib/Migrations/Version20210913225128.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        return 'Initialize DB';
+    }
+
+    public function up(Schema $schema): void
+    {
+        $this->addSql('CREATE TABLE `schedules` (id INT AUTO_INCREMENT NOT NULL, PRIMARY KEY(id))');
+        $this->addSql('
+            CREATE TABLE `rules` (
+                id INT AUTO_INCREMENT NOT NULL, 
+                rules VARCHAR(255),
```
Comment (Magomogo) [inline tests/lib/Migrations/Version20210913225128.php]: What do you plan to save in this string?

File: composer.json (hunk 2 above)
Comment (Magomogo) [reply composer.json]: IMHO we should follow this documentation: https://www.doctrine-project.org/projects/doctrine-migrations/en/3.2/reference/managing-migrations.html#executing-multiple-migrations

**Review by Magomogo — CHANGES_REQUESTED** [review]

File: composer.json (hunk 1 above)
Comment (Lehersun) [reply composer.json]: i tried to get migrations to work on php 7.3, 7.4

File: composer.json (hunk 2 above)
Comment (Lehersun) [reply composer.json]: Done

File: .travis.yml (hunk 3 above)
Comment (Lehersun) [reply .travis.yml:11]: Fixed

File: tests/lib/Migrations/Version20210913225128.php (hunk 4 above)
Comment (Lehersun) [reply tests/lib/Migrations/Version20210913225128.php]: Fixed

File: tests/lib/Migrations/Version20210913225128.php (hunk 5 above)
Comment (Lehersun) [reply tests/lib/Migrations/Version20210913225128.php]: I planned to keep the rules for the formation of the schedule

**Comment by Lehersun** [comment]
Problem 1

    - laminas/laminas-code is locked to version 4.4.2 and an update of this package was not requested.

    - laminas/laminas-code 4.4.2 requires php ^7.4 || ~8.0.0 -> your php version (7.3.27) does not satisfy that requirement.

  Problem 2

    - laminas/laminas-code 4.4.2 requires php ^7.4 || ~8.0.0 -> your php version (7.3.27) does not satisfy that requirement.

    - friendsofphp/proxy-manager-lts v1.0.5 requires laminas/laminas-code ~3.4.1|^4.0 -> satisfiable by laminas/laminas-code[4.4.2].

    - friendsofphp/proxy-manager-lts is locked to version v1.0.5 and an update of this package was not requested.

File: composer.json (hunk 1 above)
Comment (Magomogo) [reply composer.json]: Check here: https://packagist.org/packages/doctrine/migrations

Doctrine migrations package works with all PHP versions, see the version constraint: `php: ^7.2 || ^8.0`. You don't need anything to do.

File: tests/lib/Migrations/Version20210913225128.php (hunk 5 above)
Comment (Magomogo) [reply tests/lib/Migrations/Version20210913225128.php]: The most important is how to store the business rules: weekdays, dates, times etc. Do you have any suggestions?

File: composer.json (hunk 1 above)
Comment (Lehersun) [reply composer.json]: Problems with this two depending packages:
 "psr/log": "1.1.3" - psr/log 3.0 support PHP 8.0 only,
 "laminas/laminas-code": "3.4.1" - support only PHP 7.3, 7.4. Next version support only PHP 7.4, 8.0

File: tests/lib/Migrations/Version20210913225128.php (hunk 5 above)
Comment (Lehersun) [reply tests/lib/Migrations/Version20210913225128.php]: I think we need one more table 'type' to define the type of rules and store the rules in a string

File: composer.json (hunk 1 above)
Comment (Magomogo) [reply composer.json]: Why have you added dependencies `psr/log` and `laminas/laminas-code`? I don't see we directly depend on those packages.

Comment (Lehersun) [reply composer.json]: I removed dependencies but now php 8.0 has fallen off

**Review by Magomogo — COMMENTED** [review]
Looks good.

**Pull request merged**